addons/account/models/account_tax.py

```diff
@@ -2468,6 +2558,362 @@ def _prepare_tax_lines(self, base_lines, company, tax_lines=None):
+                                    By default, the base lines will be aggregated by taxes.
+        :return:                    The negative base lines representing the global discount.
+        """
+        if not base_lines:
+            return []
+
+        currency = base_lines[0]['currency_id']
+
+        # Exclude non-discountable taxes.
+        discountable_base_lines = base_lines
+        if self._has_taxes_to_exclude(base_lines):
+            discountable_base_lines = []
+            for base_line in base_lines:
+                tax_details = base_line['tax_details']
+                taxes_data = tax_details['taxes_data']
+                taxes = base_line['tax_ids'].filtered(lambda tax: tax._can_be_discounted())
+
+                if any(
+                    tax_data['tax'] not in taxes
+                    and tax_data['tax'].price_include
+                    and tax_data['tax'].include_base_amount
+                    for tax_data in taxes_data
+                ):
+                    # When the removed tax affect the base of the others, we had to remove the tax amount applied
+                    # on those taxes too.
+                    discountable_base_line = self._prepare_base_line_for_taxes_computation(
+                        base_line,
+                        price_unit=tax_details['raw_total_excluded_currency'],
+                        tax_ids=taxes,
+                        special_mode='total_excluded',
+                    )
+                    self._add_tax_details_in_base_line(discountable_base_line, company)
+                    taxes_data_for_price_unit = discountable_base_line['tax_details']['taxes_data']
+                else:
+                    taxes_data_for_price_unit = taxes_data
+
+                discountable_base_lines.append(self._prepare_base_line_for_taxes_computation(
+                    base_line,
+                    price_unit=tax_details['raw_total_excluded_currency'] + sum(
+                        tax_data['raw_tax_amount_currency']
+                        for tax_data in taxes_data_for_price_unit
+                        if tax_data['tax'].price_include and tax_data['tax'] in taxes
+                    ),
+                    tax_ids=taxes,
+                ))
+            self._add_tax_details_in_base_lines(discountable_base_lines, company)
+            self._round_base_lines_tax_details(discountable_base_lines, company)
+
+        # Compute the total discount amount to reach.
+        total_currency = self._compute_subset_base_lines_total(discountable_base_lines)
+        if amount_type == 'fixed':
+            percentage = (amount / total_currency) if total_currency else 0.0
+            target_amount_currency = amount
+        else:  # if amount_type == 'percent':
+            percentage = amount / 100.0
+            target_amount_currency = currency.round(total_currency * -percentage)
```

Comment:
```suggestion
            # Where document_total is the total of the sale order
            percentage = amount / 100.0 * document_total / total_currency
            target_amount_currency = currency.round(document_total * -amount)
```
To me this seems a better way of doing things (try to remove a certain percentage from the sale order total)
Idk if it's worth adding a test for this too

Reply:
That part is already tested in the "old" test suite about down payment. I plan to refactor them too to merge the two files in one after